docs/specs/kemsphinx.rst

```diff
@@ -30,7 +32,69 @@ has two functions:
 Therefore we must embed these KEM ciphertexts in the KEMSphinx header,
 one KEM ciphertext per mix hop.
 
-2. KEMSphinx Header Design
+2. Post Quantum Hybrid KEM
+==========================
+
+Special care must be taken in order correctly compose a hybrid post
+quantum KEM that is IND-CCA2 robust.
+
+The hybrid post quantum KEMs found in Cloudflare's circl library are
+suitable to be used with Noise or TLS but not with KEM Sphinx because
+they are not IND-CCA2 robust. Noise and TLS achieve IND-CCA2 security
+by mixing in the public keys and ciphertexts into the hash object and
+therefore do not require an IND-CCA2 KEM.
+
+Firstly, our post quantum KEM is IND-CCA2 however we must specifically take care
+to make our NIKE to KEM adapter have semantic security. Secondly, we must make a
+security preserving KEM combiner.
+
+2.1 NIKE to KEM adapter
+-----------------------
+
+We easily achieve our IND-CCA2 security by means of hashing
+together the DH shared secret along with both of the public keys:
+
+```
+func ENCAPSULATE(their_pub_key publickey) ([]byte, []byte) {
+        my_priv_key, my_pub_Key = GEN_KEYPAIR(RNG)
+        ss = DH(my_priv_key, their_pub_key)
+        ss2 = H(ss || my_pubkey || their_pubkey)
+	return my_pub_key, ss2
+}
```

Comment:
In our implementation of `core/crypto/kem/adapter/kem.go:EncapsulateDeterministically`
```go
	ss := a.nike.DeriveSecret(sk2.(*PrivateKey).privateKey, pub.publicKey)
        // ss = DH(my_priv_key=sk2 || their_pub_key=pub)
	ss2 := hashSharedSecretWithPublicKeys(a.nike.PublicKeySize(), ss, pub.publicKey.Bytes(), pk2.(*PublicKey).publicKey.Bytes())
        // ss2 = H(ss || their_pub_key=pub || my_pubkey=pk2) <--- the order of their_pub, my_pub is switched here
```
and `core/crypto/kem/adapter/kem.go:Decapsulate`:
```go
        ss := a.nike.DeriveSecret(sk.(*PrivateKey).privateKey, pk.(*PublicKey).publicKey)
	ss2 := hashSharedSecretWithPublicKeys(a.nike.PublicKeySize(), ss, sk.Public().(*PublicKey).publicKey.Bytes(), pk.(*PublicKey).publicKey.Bytes())
       // H(ss || my_pubkey || their_pubkey=pk)
```
The implementation results in matching `ss2 == shared_key`, the specification does not:
```
bob_pub , bob_ss2 = bob.ENCAPSULATE(alice_pub)
alice_ss2 = alice.DECAPSULATE(bob_pub)
// bob_ss2 = H(ss || bob_pub || alice_pub)
// alice_ss2 = H(ss || alice_pub || bob_pub)
```

Reply:
we aren't allowed to use snake case in golang. it's camel case. however i prefer to write my specs with snake case because it's easier to read.

Reply:
My suggestion is to fix the spec like this:
```diff
 func ENCAPSULATE(their_pub_key publickey) ([]byte, []byte) {
         my_priv_key, my_pub_Key = GEN_KEYPAIR(RNG)
         ss = DH(my_priv_key, their_pub_key)
-        ss2 = H(ss || my_pubkey || their_pubkey)
+        ss2 = H(ss || their_pubkey || my_pubkey)
         return my_pub_key, ss2
```

Reply:
Thanks!